src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs

```diff
@@ -94,28 +94,33 @@ public virtual IApplication Start()
                 {
                     var httpContext = contextFactory.CreateHttpContext(features);
                     httpContext.ApplicationServices = _applicationServices;
-                    var requestIdentifier = GetRequestIdentifier(httpContext);
                     contextAccessor.HttpContext = httpContext;
+                    
                     if (diagnosticSource.IsEnabled("Microsoft.AspNet.Hosting.BeginRequest"))
                     {
                         diagnosticSource.Write("Microsoft.AspNet.Hosting.BeginRequest", new { httpContext = httpContext });
                     }
-                    try
+
+                    using (logger.RequestScope(httpContext))
                     {
-                        using (logger.IsEnabled(LogLevel.Critical)
-                            ? logger.BeginScope("Request Id: {RequestId}", requestIdentifier) 
-                            : null)
+                        try
                         {
+                            logger.RequestStarting(httpContext);
                             await application(httpContext);
                         }
-                    }
-                    catch (Exception ex)
-                    {
-                        if (diagnosticSource.IsEnabled("Microsoft.AspNet.Hosting.UnhandledException"))
+                        catch (Exception ex)
                         {
-                            diagnosticSource.Write("Microsoft.AspNet.Hosting.UnhandledException", new { httpContext = httpContext, exception = ex });
+                            if (diagnosticSource.IsEnabled("Microsoft.AspNet.Hosting.UnhandledException"))
+                            {
+                                diagnosticSource.Write("Microsoft.AspNet.Hosting.UnhandledException", new { httpContext = httpContext, exception = ex });
+                            }
+                            throw;
                         }
-                        throw;
+                        finally
+                        {
+                            logger.RequestFinished(httpContext);
```

Comment:
An alternative is to hook the request completed event to have the correct status code and possibly more accurate overall time. That would need to be weighed against the overhead of the event hook per request - but it can be done only if that level is enabled - so it's no cost at all when no loggers aren't added.